### PR TITLE
[MERGE] crm, crm_iap_enrich, various: improve IAP enrich

### PR DESCRIPTION
**Pull request by pka-odoo**
PURPOSE

Cleanup automatic enrichment of leads.

SPECIFICATIONS

Right now, when we choose to automatically enrich the leads from CRM
settings, it activates a cron to periodically enrich the leads using
IAP service.

This commit improves the behavior by enriching the leads after the
records creation using cron trigger. That way it is done nearly after
creation and user gets enrich information sooner.

To make it clear to users, the description for 'auto' mode is improved to
'Enrich all leads on creation'.

Also, now we select 'auto' mode by default instead of 'manual', and
display 'Enrich' button on form view irrespective of the selected mode
(if lead meets certain conditions) unlike before. Rest of the behavior
is still same as before. For example, we still have server action which
can enrich the selected leads in batch, which is useful if there are
existing leads before we enable 'Lead Enrichment' feature.

SIDE SPECIFICATIONS

Improve lead / opportunity marketing information display on form view.

Send html toast notifications when running out of IAP credits for enrich.

Fix some corner cases for lead enrich (involving probability, email).

LINKS

Task-2269743

### CODE REVIEW

File: addons/crm_iap_lead_enrich/models/crm_lead.py [hunk 1]
```diff
@@ -29,19 +30,21 @@ def _compute_show_enrich_button(self):
...
-            ('create_date', '>', timeDelta)
-        ])
-        leads.iap_enrich(from_cron=True)
-
-    def iap_enrich(self, from_cron=False):
+    @api.model_create_multi
+    def create(self, vals_list):
+        leads = super(Lead, self).create(vals_list)
+        enrich_mode = self.env['ir.config_parameter'].get_param('crm.iap.lead.enrich.setting', 'manual')
+        if enrich_mode == 'auto' and (not self._context.get('install_mode') or self._context.get('import_file')) and not getattr(threading.currentThread(), 'testing', False):
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/models/crm_lead.py:37]: Could be great to extract right parts into a variable, to understand its purpose. Like
```
user_creation_mode = (not self._context.get('install_mode') or self._context.get('import_file')) and not getattr(threading.currentThread(), 'testing', False)
```

File: addons/crm_iap_lead_enrich/models/crm_lead.py [hunk 2]
```diff
@@ -29,19 +30,21 @@ def _compute_show_enrich_button(self):
...
+            leads_to_enrich = leads.filtered(lambda lead: lead.email_from)
+            if self._context.get('import_file'):
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/models/crm_lead.py:39]: Actually we could always apply the limit. No need to limit it with the import_file context key.

File: addons/crm_iap_lead_enrich/models/crm_lead.py [hunk 3]
```diff
@@ -29,19 +30,21 @@ def _compute_show_enrich_button(self):
...
-        ])
-        leads.iap_enrich(from_cron=True)
-
-    def iap_enrich(self, from_cron=False):
+    @api.model_create_multi
+    def create(self, vals_list):
+        leads = super(Lead, self).create(vals_list)
+        enrich_mode = self.env['ir.config_parameter'].get_param('crm.iap.lead.enrich.setting', 'manual')
+        if enrich_mode == 'auto' and (not self._context.get('install_mode') or self._context.get('import_file')) and not getattr(threading.currentThread(), 'testing', False):
+            leads_to_enrich = leads.filtered(lambda lead: lead.email_from)
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/models/crm_lead.py:38]: Maybe we could directly add "lead.probability != 100 or not lead.iap_enrich_done and lead.active" directly here.

**Review by tde-banana-odoo — COMMENTED** [review]
Validation

File: addons/crm_iap_lead_enrich/models/crm_lead.py [hunk 4]
```diff
@@ -29,19 +30,21 @@ def _compute_show_enrich_button(self):
...
-
-    def iap_enrich(self, from_cron=False):
+    @api.model_create_multi
+    def create(self, vals_list):
+        leads = super(Lead, self).create(vals_list)
+        ConfigParameter = self.env['ir.config_parameter'].sudo()
+        enrich_mode = ConfigParameter.get_param('crm.iap.lead.enrich.setting', 'manual')
+        user_creation_mode = (not self._context.get('install_mode') or self._context.get('import_file')) and not getattr(threading.currentThread(), 'testing', False)
+        if enrich_mode == 'auto' and user_creation_mode:
+            enrich_limit = ConfigParameter.get_param('crm_iap_lead_encrich_limit', 1000)
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/models/crm_lead.py:40]: Still using encrich 8-) also please use dots to be coherent through all CRM config parameters use.

File: addons/crm_iap_lead_enrich/__init__.py [hunk 5]
```diff
@@ -2,13 +2,3 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from . import models
-
-from odoo.api import Environment, SUPERUSER_ID
-
-
-def _synchronize_cron(cr, registry):
```
Comment (awa-odoo) [inline addons/crm_iap_lead_enrich/__init__.py:9]: Why is this removed exactly?

One could still set the parameter to "manual" before installing this module, correct?

File: addons/crm_iap_lead_enrich/models/crm_lead.py [hunk 6]
```diff
@@ -53,7 +62,7 @@ def iap_enrich(self, from_cron=False):
                         {'lead_ids': tuple(leads.ids)}, log_exceptions=False)
                     for lead in leads:
                         # If lead is lost, active == False, but is anyway removed from the search in the cron.
-                        if lead.probability == 100 or lead.iap_enrich_done:
+                        if lead.probability == 100 or lead.iap_enrich_done and not lead.active:
```
Comment (awa-odoo) [inline addons/crm_iap_lead_enrich/models/crm_lead.py:61]: Not sure I understand why this change is related to your commit here?

File: addons/crm_iap_lead_enrich/models/res_config_settings.py [hunk 7]
```diff
@@ -1,21 +0,0 @@
...
-        values = super(ResConfigSettings, self).get_values()
-        cron = self.sudo().with_context(active_test=False).env.ref('crm_iap_lead_enrich.ir_cron_lead_enrichment', raise_if_not_found=False)
-        values['lead_enrich_auto'] = 'auto' if cron and cron.active else 'manual'
-        return values
-
-    def set_values(self):
-        super(ResConfigSettings, self).set_values()
-        cron = self.sudo().with_context(active_test=False).env.ref('crm_iap_lead_enrich.ir_cron_lead_enrichment', raise_if_not_found=False)
-        if cron:
-            cron.active = self.lead_enrich_auto == 'auto'
```
Comment (awa-odoo) [inline addons/crm_iap_lead_enrich/models/res_config_settings.py:21]: Again, not sure why the setting and the cron are de-synchronized now.

File: addons/crm_iap_lead_enrich/models/crm_lead.py [hunk 8]
```diff
@@ -19,10 +19,6 @@ class Lead(models.Model):
 
     @api.depends('email_from', 'probability', 'iap_enrich_done', 'reveal_id')
     def _compute_show_enrich_button(self):
-        config = self.env['ir.config_parameter'].sudo().get_param('crm.iap.lead.enrich.setting', 'manual')
-        if not config or config != 'manual':
-            self.show_enrich_button = False
-            return
```
Comment (awa-odoo) [inline addons/crm_iap_lead_enrich/models/crm_lead.py:25]: Looks unrelated to this commit.

File: addons/crm_iap_lead_enrich/models/crm_lead.py [hunk 9]
```diff
@@ -87,14 +83,17 @@ def iap_enrich(self, from_cron=False):
                             iap_response = self.env['iap.enrich.api']._request_enrich(lead_emails)
                         except iap_tools.InsufficientCreditError:
                             _logger.info('Sent batch %s enrich requests: failed because of credit', len(lead_emails))
-                            if not from_cron:
+                            if not from_cron and not self._context.get('import_file'):
```
Comment (awa-odoo) [inline addons/crm_iap_lead_enrich/models/crm_lead.py:86]: Looks unrelated to this commit.

File: addons/crm_iap_lead_enrich/models/crm_lead.py [hunk 10]
```diff
@@ -92,6 +92,7 @@ def iap_enrich(self, from_cron=False):
                                     {
                                         'type': 'simple_notification',
                                         'message': self.env.ref('crm_iap_lead_enrich.mail_message_lead_enrich_no_credit')._render(data),
+                                        'message_is_html': True,
```
Comment (awa-odoo) [inline addons/crm_iap_lead_enrich/models/crm_lead.py:95]: Meaning your previous commit does not work?

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js [hunk 11]
```diff
@@ -397,6 +397,7 @@ function factory(dependencies) {
                 this.env.services['notification'].notify({
                     message: data.message,
                     sticky: data.sticky,
+                    messageIsHtml: data.message_is_html,
```
Comment (awa-odoo) [inline addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:400]: Looks a bit unrelated to your task overall.

**Review by awa-odoo — COMMENTED** [review]
Hello @nla-odoo 

Overall a lot of changes, where we should only have a cron trigger in the crm.lead create method (or at least that's the main goal).
Not sure how all those are related to the task purpose, probably to double-check.

Thanks for your work.

**Comment by awa-odoo** [comment]
Also, there seems to be a lot of file conflicts, could you rebase?

File: addons/crm_iap_lead_enrich/models/crm_lead.py (hunk 6 above)
Comment (nla-odoo) [reply addons/crm_iap_lead_enrich/models/crm_lead.py:61]: Hello @awa-odoo 

Actually, we don't want to show the Bus notification if a user is importing the file.
Because it may trigger this notification multiple times on the user screen.

Thanks

File: addons/crm_iap_lead_enrich/models/res_config_settings.py (hunk 7 above)
Comment (nla-odoo) [reply addons/crm_iap_lead_enrich/models/res_config_settings.py:21]: Done

File: addons/crm_iap_lead_enrich/models/crm_lead.py (hunk 9 above)
Comment (nla-odoo) [reply addons/crm_iap_lead_enrich/models/crm_lead.py:86]: Hello @awa-odoo 

Actually, we don't want to show the Bus notification if a user is importing the file.
Because it may trigger this notification multiple times on the user screen.

Thanks

File: addons/crm_iap_lead_enrich/models/crm_lead.py (hunk 10 above)
Comment (nla-odoo) [reply addons/crm_iap_lead_enrich/models/crm_lead.py:95]: It was working but showing escaped data instead of HTML note

File: addons/crm_iap_lead_enrich/models/crm_lead.py (hunk 6 above)
Comment (nla-odoo) [reply addons/crm_iap_lead_enrich/models/crm_lead.py:61]: We don't want to enrich the archived lead.

File: addons/crm_iap_lead_enrich/models/crm_lead.py (hunk 9 above)
Comment (awa-odoo) [reply addons/crm_iap_lead_enrich/models/crm_lead.py]: Hmmm.

To double check but I'm pretty sure the CRON will NOT share the same context as the import process anyway.

Comment (nla-odoo) [reply addons/crm_iap_lead_enrich/models/crm_lead.py]: Hello @awa-odoo 

Yes, you are right. Cron trigger is not sharing context.

File: addons/crm/data/crm_lead_prediction_data.xml [hunk 12]
```diff
@@ -23,6 +23,10 @@
         <record id="frequency_field_tag_ids" model="crm.lead.scoring.frequency.field">
             <field name="field_id" ref="crm.field_crm_lead__tag_ids"/>
         </record>
+        <record id="crm_iap_lead_enrich_setting" model="ir.config_parameter">
+            <field name="key">crm.iap.lead.enrich.setting</field>
+            <field name="value">auto</field>
+        </record>
```
Comment (awa-odoo) [inline addons/crm/data/crm_lead_prediction_data.xml:29]: Instead we could remove this data and consider the default to be "auto" when calling "get_param".

So instead of 
`enrich_mode = self.env['ir.config_parameter'].sudo().get_param('crm.iap.lead.enrich.setting', 'manual')`

You would have
`enrich_mode = self.env['ir.config_parameter'].sudo().get_param('crm.iap.lead.enrich.setting', 'auto')`

Isn't it more simple?

File: addons/crm_iap_enrich/models/crm_lead.py [hunk 13]
```diff
@@ -41,6 +37,15 @@ def _iap_enrich_leads_cron(self):
         ])
         leads.iap_enrich(from_cron=True)
 
+    @api.model_create_multi
+    def create(self, vals_list):
+        leads = super(Lead, self).create(vals_list)
+        enrich_mode = self.env['ir.config_parameter'].sudo().get_param('crm.iap.lead.enrich.setting', 'manual')
+        if enrich_mode == 'auto':
+            cron = self.env.ref('crm_iap_enrich.ir_cron_lead_enrichment')
```
Comment (awa-odoo) [inline addons/crm_iap_enrich/models/crm_lead.py:45]: Could add "raise_if_not_found=False" and check that cron exists before calling "_trigger" on it.
Would make it safer if the cron is removed for some reason.

File: addons/crm_iap_enrich/models/crm_lead.py [hunk 14]
```diff
@@ -41,6 +37,15 @@ def _iap_enrich_leads_cron(self):
...
+            cron._trigger(fields.Datetime.now())
```
Comment (awa-odoo) [inline addons/crm_iap_enrich/models/crm_lead.py:46]: Don't think you have to pass "fields.Datetime.now()", I think the method does that if you don't pass it anything.

**Review by awa-odoo — COMMENTED** [review]
Hello @nla-odoo 

Looking much better / cleaner.

I left a few more polishing comments, but we are close!

Thanks for your work.

File: addons/crm_iap_enrich/models/crm_lead.py (hunk 14 above)
Comment (nla-odoo) [reply addons/crm_iap_enrich/models/crm_lead.py]: Hello @awa-odoo 

Changes done.

Thanks

File: addons/crm_iap_enrich/models/crm_lead.py [hunk 15]
```diff
@@ -53,7 +59,7 @@ def iap_enrich(self, from_cron=False):
                         {'lead_ids': tuple(leads.ids)}, log_exceptions=False)
                     for lead in leads:
                         # If lead is lost, active == False, but is anyway removed from the search in the cron.
-                        if lead.probability == 100 or lead.iap_enrich_done:
+                        if lead.probability == 100 or lead.iap_enrich_done and not lead.active:
```
Comment (awa-odoo) [inline addons/crm_iap_enrich/models/crm_lead.py:62]: Somehow this change is back.

Can you either remove it or explain why you do it?
And I mean, where it was asked in the task specifications?
I don't see any reference to the archive mechanism in there.

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge
